### ISSUE TITLE
Fowlkes–Mallows: prevent int overflow in denominator (supervised.py#fowlkes_mallows_score)

### ISSUE THREAD

**Issue by emerson-gray** [agyn-sandbox/scikit-learn#6]
Summary
- File: sklearn/metrics/cluster/supervised.py
- Function: fowlkes_mallows_score
- Current line (circa L859):
  return tk / np.sqrt(pk * qk) if tk != 0. else 0.
- Problem: RuntimeWarning: overflow encountered in int_scalars when (pk * qk) overflows int32, producing NaN. We need a robust float64 computation of the denominator that avoids integer overflow, preserves existing behavior for tk==0, and maintains output type as float.

Recommended fix approach
Option A (recommended: minimal, stable, fast)
- Compute the denominator in float64: denom = np.sqrt(np.float64(pk) * np.float64(qk)).
- Return 0.0 early if tk == 0 (preserves current semantics and also covers pk==0 or qk==0 which imply tk==0).
- Rationale: Casting to float64 before multiplication eliminates integer overflow. For realistic dataset sizes, float64 range handily covers pk*qk (O(n^4) worst bound is not reached in practice for n in the millions; even if extremely large, float64 returns inf which yields 0 with tk==0, and otherwise a finite ratio as long as denom < inf). This approach keeps a single multiply and a single sqrt; maintains backward compatibility with negligible overhead.

Option B (alternative): algebraic rewrite
- Compute: fm = np.sqrt(np.float64(tk) / np.float64(pk)) * np.sqrt(np.float64(tk) / np.float64(qk)).
- Pros: Avoids forming pk*qk; might be marginally more stable for extremely large magnitudes.
- Cons: Requires two divisions and two sqrts; requires explicit guards for pk==0 and qk==0 to avoid division-by-zero; slightly more branchy and less minimal than Option A. Numerically, both options are stable over our expected domain once float64-cast is applied.

Chosen solution: Option A
- Minimal, preserves current structure and branch (tk == 0 → 0.0), and avoids extra divisions. No behavior changes except the removal of overflow warnings and preventing NaN from int overflow.

Zero and edge-case behavior
- tk == 0: return 0.0 (unchanged). This covers disjoint clusterings and also cases where pk==0 or qk==0 (which imply no positive pairs, hence tk==0).
- pk == 0 or qk == 0: tk must be 0 for valid clusterings; thus the existing tk==0 guard returns 0.0. Optionally, for clarity, we can also add an explicit early return 0.0 if pk == 0 or qk == 0, but it is not required for correctness.
- Output range remains [0.0, 1.0].

Dtype handling
- Inputs: labels may be Python int, numpy int32/64.
- Internals: Ensure denominator computation casts pk and qk to np.float64 before multiplication and sqrt.
- Output: always return a Python float backed by float64 (np.float64), consistent with scikit-learn metrics.

Performance considerations
- Option A introduces two scalar float64 casts and performs one multiply and one sqrt: negligible overhead compared to current code.
- No additional allocations or loops.

Code diff sketch (surgical change only)
In sklearn/metrics/cluster/supervised.py inside fowlkes_mallows_score, replace the final return with:

    # existing early exit remains
    if tk == 0:
        return 0.0

    # robust float64 denominator to avoid integer overflow
    denom = np.sqrt(np.float64(pk) * np.float64(qk))
    return np.float64(tk) / denom

Or, if preserving the single-line style, minimally:

    return (np.float64(tk) / np.sqrt(np.float64(pk) * np.float64(qk))) if tk != 0 else 0.0

Test plan
1) Basic correctness
- Identical labels
  y_true = [0, 0, 1, 1, 2, 2]
  y_pred = [0, 0, 1, 1, 2, 2]
  Expect fowlkes_mallows_score(...) == 1.0

- Completely different clusters
  y_true = [0, 0, 0, 1, 1, 1]
  y_pred = [0, 1, 2, 3, 4, 5]
  Expect 0.0

2) Zero-pairs edge cases
- All singletons vs all singletons
  y_true = [0, 1, 2, 3]
  y_pred = [10, 11, 12, 13]
  Expect 0.0 (consistent with current behavior tk==0)

- One cluster vs singletons (no shared pairs)
  y_true = [0, 0, 0, 0]
  y_pred = [0, 1, 2, 3]
  Expect 0.0

3) Large counts triggering previous overflow
- Construct labels so that pk and qk exceed int32 range, but tk > 0.
  For example, repeated labels with large class sizes:

  import numpy as np
  from sklearn.metrics.cluster import fowlkes_mallows_score

  n = 65536  # 65536 choose 2 ≈ 2.147e9 (> 2^31-1)
  y_true = np.zeros(n, dtype=np.int32)
  y_pred = np.zeros(n, dtype=np.int32)

  # Before fix (current main): may emit RuntimeWarning and potentially NaN on some platforms
  # After fix (this branch): no warnings; result should be 1.0
  fm = fowlkes_mallows_score(y_true, y_pred)
  assert np.isfinite(fm)
  assert abs(fm - 1.0) < 1e-12

- Variant: two large equal-sized clusters
  n = 2 * 50000
  y_true = np.repeat([0, 1], [50000, 50000]).astype(np.int32)
  y_pred = y_true.copy()
  fm = fowlkes_mallows_score(y_true, y_pred)
  Expect ~1.0 with no warnings

4) Dtype coverage
- Use Python lists, np.int32 arrays, and np.int64 arrays for labels in the above tests to ensure consistent float64 outputs.

5) No warnings
- Wrap calls with warnings.catch_warnings(record=True) as w: and assert w contains no overflow warnings.

Reproduction of current overflow (before fix)
This demonstrates the overflow symptom originating from integer multiplication of pk and qk (mirroring the current line):

  import numpy as np
  from sklearn.metrics.cluster import fowlkes_mallows_score
  import warnings

  warnings.simplefilter('always')

  n = 65536
  y_true = np.zeros(n, dtype=np.int32)
  y_pred = np.zeros(n, dtype=np.int32)

  fm = fowlkes_mallows_score(y_true, y_pred)
  print('FM=', fm)

Observed (before fix) on affected builds:
- RuntimeWarning: overflow encountered in int_scalars
    return tk / np.sqrt(pk * qk) if tk != 0. else 0.
- Output may be NaN on some platforms due to overflow in pk*qk before sqrt.

Backward compatibility
- Returned values for normal ranges are unchanged up to floating-point round-off.
- Behavior for tk==0 remains 0.0 (consistent with current code and scikit-learn’s documented range and examples).
- No API changes or new dependencies.

Acceptance criteria
- No overflow warnings from fowlkes_mallows_score on large-label test cases.
- Output is finite float64 for all tested dtypes and edge cases.
- Existing tests pass; new tests described above are added.

Code owners: metrics/cluster maintainers.
